### PR TITLE
Adds Local OOC ported from cit, tested and fully functional

### PR DESCRIPTION
**Pull request by Bampotio**
Does what it says on the tin reall

## About The Pull Request
Adds LOOC

## Why It's Good For The Game
Lets people ask question and stuff without shitting up OOC

## Changelog
:cl:
add: LOOC
/:cl:

### CODE REVIEW

**Comment by JohnFulpWillard** [comment]
If people have a question about the game mechanics, they can ask in a Mentorhelp.
If people want to ask someone about something, they can do it IC.

1) this feels like it’s just removing the need for people to talk to eachother IC when they can just do it OOC.
2) I don’t see any ocassion where LOOC can be used properly due to wha I stated above.
3) We barely have any admins as it is and the team relies on grief patrols as a result, now you want to add a completely new OOC channel, and one that can’t be seen by the people writing the grief patrols.
4) Your ‘Why it’s good for the game’ is usually supposed to have an actual reason as to why it’s good for the game, and not just a description of Mentorhelp.

**Comment by Bampotio** [comment]
Have you actually played on a server with LOOC its the only reason i made this, its fucking quality.  People can still mhelp, and people can still ask things IC this just adds another option that doesn't require telling the entire server something

File: fulp_modules/LOOC_port/LOOC.dm [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+			log_admin("[key_name(src)] has attempted to advertise in LOOC: [msg]")
+			return
+		if(mob.stat)
+			to_chat(src, "<span class='danger'>You cannot use LOOC while unconscious or dead.</span>")
+			return
+		if(isdead(mob))
+			to_chat(src, "<span class='danger'>You cannot use LOOC while ghosting.</span>")
+			return
+
+
```
Comment (urococt) [inline fulp_modules/LOOC_port/LOOC.dm:45]: [Copy-pasted](https://github.com/fulpstation/fulpstation/blob/master/code/modules/client/verbs/ooc.dm#L25)

```suggestion
	if(is_banned_from(ckey, "OOC"))
		to_chat(src, "<span class='danger'>You have been banned from OOC.</span>")
		return
	if(QDELETED(src))
		return
```

File: fulp_modules/LOOC_port/LOOC.dm [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+			else
+				to_chat(C, "<span class='looc'>[ADMIN_FLW(usr)] <span class='prefix'>[prefix]:</span> <EM>[src.key]/[src.mob.name]:</EM> <span class='message'>[msg]</span></span>")
+
+/mob/proc/get_top_level_mob()
+    if(istype(src.loc,/mob)&&src.loc!=src)
+        var/mob/M=src.loc
+        return M.get_top_level_mob()
+    return src
+
+proc/get_top_level_mob(var/mob/S)
```
Comment (urococt) [inline fulp_modules/LOOC_port/LOOC.dm:84]: [Copy-pasted](https://github.com/Citadel-Station-13/Citadel-Station-13/blob/9f50ea66c80485af8dc0975a7bf8b3f32ea41cf1/code/modules/mob/say_vr.dm#L11)

```suggestion
/proc/get_top_level_mob(mob/S)
```

Why is this necessary though?

File: fulp_modules/LOOC_port/LOOC.dm [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+			to_chat(src, "<span class='danger'>You cannot use LOOC while unconscious or dead.</span>")
+			return
+		if(isdead(mob))
+			to_chat(src, "<span class='danger'>You cannot use LOOC while ghosting.</span>")
+			return
+
+
+
+	msg = emoji_parse(msg)
+
```
Comment (urococt) [inline fulp_modules/LOOC_port/LOOC.dm:48]: [Copy-pasted](https://github.com/fulpstation/fulpstation/blob/67d27e3d83e82ba022d0cfef97d84c8822174fed/code/modules/client/verbs/ooc.dm#L39)

```suggestion

	if(SSticker.HasRoundStarted() && (msg[1] in list(".",";",":","#") || findtext_char(msg, "say", 1, 5)))
		if(alert("Your message \"[raw_msg]\" looks like it was meant for in game communication, say it in LOOC?", "Meant for LOOC?", "Yes", "No") != "Yes")
			return

```

File: fulp_modules/LOOC_port/LOOC.dm [hunk 4]
```diff
@@ -0,0 +1,88 @@
...
+	var/list/heard = get_hearers_in_view(7, get_top_level_mob(src.mob))
+	for(var/mob/M in heard)
+		if(!M.client)
+			continue
+		var/client/C = M.client
+		if (C in GLOB.admins)
+			continue //they are handled after that
+
+		if (isobserver(M))
+			continue //Also handled later.
```
Comment (urococt) [inline fulp_modules/LOOC_port/LOOC.dm:60]: Except they aren't.

**Review by urococt — COMMENTED** [review]
This will hopefully reduce OOC in IC, which is good for a beginner server with MRP aspirations. I can see this being very helpful for teaching new players in medbay.

Showing all LOOC messages to ghosts would help to prevent abuse. I don't know what the best way to handle that would be. Adding a preference for ghost LOOC might be overkill, and lumping LOOC in with OOC might not be granular enough.

I don't think abuse is much of a concern though. As you said, this doesn't seem to enable rule-breaking behavior that can't be accomplished with IC, OOC, or emotes.

File: fulp_modules/LOOC_port/LOOC.dm (hunk 4 above)
Comment (Bampotio) [reply fulp_modules/LOOC_port/LOOC.dm:67]: yeh i just copied most of this from citadel and edited some stuff so it would compile and work

File: fulp_modules/LOOC_port/LOOC.dm [hunk 5]
```diff
@@ -0,0 +1,95 @@
...
+	if(is_banned_from(ckey, "OOC"))
+		to_chat(src, "<span class='danger'>You have been banned from OOC.</span>")
+		return
+	if(QDELETED(src))
+		return
+
+	msg = emoji_parse(msg)
+
+	if(SSticker.HasRoundStarted() && (msg[1] in list(".",";",":","#") || findtext_char(msg, "say", 1, 5)))
+		if(alert("Your message \"[raw_msg]\" looks like it was meant for in game communication, say it in LOOC?", "Meant for LOOC?", "Yes", "No") != "Yes")
```
Comment (urococt) [inline fulp_modules/LOOC_port/LOOC.dm:53]: I am dumb.

```suggestion
		if(alert("Your message \"[msg]\" looks like it was meant for in game communication, say it in LOOC?", "Meant for LOOC?", "Yes", "No") != "Yes")
```

File: fulp_modules/LOOC_port/LOOC.dm [hunk 6]
```diff
@@ -0,0 +1,92 @@
...
+		to_chat(src, "<span class='danger'>You have been banned from OOC.</span>")
+		return
+	if(QDELETED(src))
+		return
+
+	msg = emoji_parse(msg)
+
+	mob.log_talk(msg,LOG_OOC, tag="LOOC")
+
+	var/list/heard = get_hearers_in_view(7, get_top_level_mob(src.mob))
```
Comment (urococt) [inline fulp_modules/LOOC_port/LOOC.dm:54]:
```suggestion
	var/list/heard = get_hearers_in_view(7, src.mob.get_top_level_mob())
```

And delete the proc that the linter is complaining about.

File: fulp_modules/LOOC_port/LOOC.dm [hunk 7]
```diff
@@ -0,0 +1,90 @@
...
+		if(C.prefs.chat_toggles & CHAT_OOC)
+			var/prefix = "(R)LOOC"
+			if (C.mob in heard)
+				prefix = "LOOC"
+			if(GLOB.LOOC_COLOR)
+				to_chat(C, "<font color='[GLOB.LOOC_COLOR]'><b>[ADMIN_FLW(usr)] <span class='prefix'>[prefix]:</span> <EM>[src.key]/[src.mob.name]:</EM> <span class='message'>[msg]</span></b></font>")
+			else
+				to_chat(C, "<span class='looc'>[ADMIN_FLW(usr)] <span class='prefix'>[prefix]:</span> <EM>[src.key]/[src.mob.name]:</EM> <span class='message'>[msg]</span></span>")
+
+
```
Comment (urococt) [inline fulp_modules/LOOC_port/LOOC.dm:85]: This patch will enable ghosts to see LOOC in view range when OOC is enabled, and all LOOC messages when Ghost Whispers and OOC are enabled.

I actually tested this time.

```suggestion
	for(var/mob/M in GLOB.player_list)
		if(QDELETED(M))
			continue
		if(!isobserver(M))
			continue
		if(M.client in GLOB.admins)
			continue //Already handled
		if(!(M.client.prefs.chat_toggles & CHAT_OOC))
			continue
		if((!(M.client.prefs.chat_toggles & CHAT_GHOSTWHISPER)) && !(M in heard))
			continue //Piggyback on the ghost whispers preference to show/hide distant LOOC for ghosts
		
		var/link = FOLLOW_LINK(M, mob)
		if(GLOB.LOOC_COLOR)
			to_chat(M.client, "[link] <font color='[GLOB.LOOC_COLOR]'><b><span class='prefix'>LOOC:</span> <EM>[src.mob.name]:</EM> <span class='message'>[msg]</span></b></font>")
		else
			to_chat(M.client, "[link] <span class='looc'><span class='prefix'>LOOC:</span> <EM>[src.mob.name]:</EM> <span class='message'>[msg]</span></span>")

```

Comment (Bampotio) [reply fulp_modules/LOOC_port/LOOC.dm:85]: cool im not 100% on ghosts being able to hear all looc though, it seems unnecessary

Comment (urococt) [reply fulp_modules/LOOC_port/LOOC.dm:85]: You might want to get rid of the observer check on L66 then, so ghosts can hear LOOC in view.

Comment (ATH1909) [reply fulp_modules/LOOC_port/LOOC.dm:85]: > cool im not 100% on ghosts being able to hear all looc though, it seems unnecessary

eh, sometimes there's some good banter or extra layers of story in there.

plus, the more people who can see and report LOOC abuse, the better, IMO.

Comment (Bampotio) [reply fulp_modules/LOOC_port/LOOC.dm:85]: guess so il commit it and if it gets merge and becomes and issue i can always revert it

**Comment by Helianthus00** [comment]
With mentors, we don't need a feature like this, and it just adds another layer of things to be moderated. Closing.
